### ISSUE TITLE
Extension keeps building code, never finishing

### ISSUE THREAD

**Issue by fo-fo** [redhat-developer/vscode-java#1341]
I'm trying to use the vscode-java extension in VSCode, but it keeps on building the same source code (it seems) again and again, hogging 100% CPU.

I see lines like this in the `Task - Java Build Status` Terminal: `1e7bf092 Building: 72% Compiling foo/src/test/java/fi/bar/xyzzy/controller [720/1000]`. After it completes, it starts again.

I set the "Trace" setting in Settings to "messages", and saw this in the logs multiple times: `[Trace - 9:47:08 AM] Sending notification 'workspace/didChangeWatchedFiles'.`, followed by a bunch of `[Trace - 9:47:08 AM] Received notification 'language/progressReport'.`

My guess is that it somehow thinks that some of the watched files have changed. **How can I find out which files are causing the build to re-trigger?**

##### Environment
- Operating System: macOS Catalina 10.15.3
- JDK version: jdk1.8.0_161
- Visual Studio Code version: 1.43.2
- Java extension version: 0.58.0

##### Steps To Reproduce

Unfortunately it only happens in our fairly large proprietary project.

##### Current Result

Keeps rebuilding the code forever.

##### Expected Result

Should stop building after a reasonable time.

**Comment by testforstephen** [comment]
@fo-fo if you change the java trace setting to "verbose", you will see the full logs about "workspace/didChangeWatchedFiles" notification.

**Comment by fo-fo** [comment]
@testforstephen Thanks, I was able to narrow down the cause. It seems that RAML code generation is the culprit, as didChangeWatchedFiles refers to the generated files and I can see the generated files' time stamps changing every few minutes. So, it seems like the build changes the files, which then triggers didChangeWatchedFiles and a rebuild, which again changes the files, and so on.

I'm not sure how to fix it though. We're using `springmvc-raml-plugin` Maven plugin and a custom Ant buildfile for preprocessing the RAML sources, as well as an Ant Replace task for post processing the generated files. (By the way, the issue doesn't happen in IntelliJ IDEA.)

Any ideas where I should start looking?

**Comment by testforstephen** [comment]
The built-in builder won't support ant, and it's an incremental builder, which is supposed to not regenerate the files. Do you have other building jobs running in the background?

**Comment by fo-fo** [comment]
@testforstephen Nope, no other build jobs running. And as soon as I turn off auto build I can see from the RAML file timestamps that the files are no longer being regenerated. Moreover, if I disable `springmvc-raml-plugin` in Maven's `pom.xml` the issue disappears. If I disable the Ant jobs but leave `springmvc-raml-plugin` enabled, the issue persists, so it seems like the Ant jobs are not the cause.

We also have `swagger-codegen-maven-plugin` and `openapi-generator-maven-plugin`, they work without a problem, so it must be something specific to how `springmvc-raml-plugin` works.

**Comment by testforstephen** [comment]
OK. By default, Intellij IDEA doesn't compile classes automatically, that's probably the reason why you didn't get the issue there.

**Comment by ianldgs** [comment]
This causes problems if you try to run maven commands. 
Some files will be locked or modified by the autobuild process and then maven will fail.
Ideally, the autobuild should only run if a file is changed, is that possible?

**Comment by cammarb** [comment]
I've been experiencing a similar issue. When opening a java project, the java build status is stuck at 68% Refreshing. Also, the extension Java Project and Spring Boot Dashboard never load. 

![image](https://user-images.githubusercontent.com/44549364/126879797-470d6dda-d8fc-4195-bd3e-4ae45f7f485f.png)

I have googled and I haven't found a solution.

**Comment by snjeza** [comment]
@cammarb Could you check https://github.com/redhat-developer/vscode-java/issues/2020#issuecomment-880272712

**Comment by ianmarxent** [comment]
I had a similar issue; stuck building and would never finish. I checked out the troubleshooting guide [here](https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting) and found that cleaning out my workspace directory  solved the problem for me (steps from link pasted below):

- Open the command palette
- select Java: Clean the Java Language Server Workspace
- select Restart and delete from the confirmation prompt

**Comment by SlappyAUS** [comment]
Has anyone been able to solve this? I am running since adding WSDL code gen though maven plugin cxf-codegen-plugin, VSCode won stop iterating over "Building" -> "Refreshing Workspace" in the Java Build Status Console. Very high temperatures and CPU usage.

**Comment by ralphavalon** [comment]
@SlappyAUS so, what @snjeza posted actually solves it. It just sucks that, until today, I still need to do it. I wish this was just fixed already.

**Comment by jterry-viacomcbs** [comment]
@ianmarxent bless you, this has saved me so much work, i've been blowing away these things by hand for months

**Comment by rafaelceravolo** [comment]
> @fo-fo if you change the java trace setting to "verbose", you will see the full logs about "workspace/didChangeWatchedFiles" notification.

@fo-fo or @testforstephen I have the same problem (code keeps recompiling). Following your suggestion,  I found the configuration to set to "verbose", but didn't find where the logs are logged. 

Where can I see them?

**Comment by dhruv1233** [comment]
i also face the same problem

**Comment by snjeza** [comment]
@dhruv1233 could you, please, provide a project example reproducing the error?

**Comment by AntonioB-ThoughKnee** [comment]
This problem may be because java process extension may be killed by another extension, my problem was seeing java extension doing its thing starting the projects normally and seconds later just stopped and getting stuck in an infinite loop trying to run.
My solution was disbaling ALL extensions except java debugger and java redhat extension, restarting vs code and then I saw the java finally working in vs code, then re enabled all the extensions and I was good to go

Hope it works if you're having that problem

**Comment by alahaouas** [comment]
I tried the F1 "Java clean ..." action but the issue is still occurring.

I make no changes to the Java code or to any file, just browsing the project. 
And I notice in the "Java Build Status" terminal that building/validation/publishing/registering/etc... goes in a loop.
Then, I see via htop command that the Redhat extension uses lots of CPU.

**Comment by RollerKing** [comment]
me too, I just open https://github.com/apache/pulsar on wsl using vscode server, and  "Java Build Status" is repeating in a loop

**Comment by CMiksche** [comment]
@snjeza This issue is not solved at all. I also experience this problem.

**Comment by snjeza** [comment]
@CMiksche Could you attach a project examples reproducing the issue?
What extensions do you use?

**Comment by snjeza** [comment]
I have tried the following:

```
- remove the Gradle For Java extension. See https://github.com/microsoft/vscode-gradle/issues/1441#issuecomment-2683277597
$ rm -rf ~/.vscode/extensions/vscjava.vscode-gradle-*
$ rm -rf ~/.config/Code/User/globalStorage/redhat.java/
$ git clone git@github.com:apache/pulsar.git
$ cd pulsar
$ ./mvnw clean verify -DskipTests
$ code .  
- wait for the build to finish; it takes about 5 minutes on my machine
$ reload the VS Code window
- it takes a few seconds on my machine
```
cc @rgrunber

**Comment by iwangbowen** [comment]
It happened to me multiple times. I tried `rebuild projects` and `clean java language workspace` and it still didn't work. I have no choice but to delete the whole project and re-clone it to make it work.

**Comment by snjeza** [comment]
@iwangbowen Do you use Gradle for Java?

**Comment by rgrunber** [comment]
> I have tried the following:

With the instructions, on a completely fresh import (and your instructions), I see :

```
"time.projectsinitialized": "59718",
"time.serviceready": "59966",
"time.buildFinished": "64211",
```

So it takes about a minute to build. The remaining 5 minutes are spent on indexing ~2500 library files. Afterwards, subsequent reloads take around 40s for the build to complete for me.

```
"time.projectsinitialized": "2843",
"time.serviceready": "3048",
"time.buildFinished": "39064"
```

Definitely not a few seconds for everything to complete, but the language server would be responsive in a few seconds based on "service ready" time.

Also, I noticed pulsar, whether building on cli or with our extension triggers the native passphrase dialog to unlock a private key. However, if the issue is a build loop, I don't think it would be related, since the dialog would only freeze the build in place.

Does anyone have a sample project where the problem is occurring with any consistency? This issue was opened 5 years ago, and has been closed for 3, so it's likely the original data isn't going to help here.

**Comment by rgrunber** [comment]
Does anyone have a sample project that demonstrates the build never finishing ?

**Comment by rgrunber** [comment]
If this is reproducible, would you be able to set `"java.trace.server": "verbose"` in your settings, reproduce the issue, and then go to `Java: Open All Log Files` from the command palette, and then send the contents of the `.log` and `client.log.YYYY-MM-DD` from around when the problem occured. This might give a better idea of what's going wrong.

**Comment by JackToner2025** [comment]
[.log](https://github.com/user-attachments/files/19720258/default.log)

I got an error trying to upload client log ending in .11

The code compiles without error in maven and I re-installed VS Code but the problem persists.

**Comment by snjeza** [comment]
> I got an error trying to upload client log ending in .11
The code compiles without error in maven and I re-installed VS Code but the problem persists.

@JackToner2025 Could you try VS Code pre-release? See https://github.com/redhat-developer/vscode-java/issues/4020#issuecomment-2783147767
